### PR TITLE
ci: add integration test for external python script

### PR DESCRIPTION
**Pull request by leseb**
**Description of your changes:**

The CI runs a github action that builds and deploys a Rook cluster and
then execute various tests on the create-external-cluster-resources.py
script.

Signed-off-by: Sébastien Han <seb@redhat.com>

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

[skip ci]

### CODE REVIEW

**Comment by galexrt** [comment]
@leseb Minikube will not run with a VM due to GitHub actions not having nested virtualization enabled due to VMs used for the actions runners, see https://github.com/actions/virtual-environments/issues/183#issuecomment-675026289

**Comment by leseb** [comment]
> @leseb Minikube will not run with a VM due to GitHub actions not having nested virtualization enabled due to VMs used for the actions runners, see [actions/virtual-environments#183 (comment)](https://github.com/actions/virtual-environments/issues/183#issuecomment-675026289)

I think this is using the docker driver so we don't do nested virt.

File: Makefile [hunk 1]
```diff
@@ -95,7 +95,8 @@ build.version:
 	@mkdir -p $(OUTPUT_DIR)
 	@echo "$(VERSION)" > $(OUTPUT_DIR)/version
 
-build.common: build.version helm.build mod.check
+# build.common: build.version helm.build mod.check
```
Comment (leseb) [inline Makefile:98]: helm fails with `Error: Invalid Semantic Version` perhaps because we run on dirty checkout?

File: pkg/daemon/ceph/osd/daemon.go [hunk 2]
```diff
@@ -203,13 +203,13 @@ func Provision(context *clusterd.Context, agent *OsdAgent, crushLocation string)
...
 	// Check for the presence of LVM on the host when NOT running on PVC
 	// since this scenario is still using LVM
-	if !agent.pvcBacked {
-		ne := NewNsenter(context, lvmCommandToCheck, []string{"--help"})
-		err := ne.checkIfBinaryExistsOnHost()
-		if err != nil {
-			return errors.Wrapf(err, "binary %q does not exist on the host, make sure lvm2 package is installed", lvmCommandToCheck)
-		}
-	}
+	// if !agent.pvcBacked {
```
Comment (leseb) [inline pkg/daemon/ceph/osd/daemon.go:206]: Somehow the prepare job cannot find the binary... Might be due to the docker driver of minikube...

**Comment by satoru-takeuchi** [comment]
@leseb Is CI trying to migrate from Jenkins to github actions?

**Comment by leseb** [comment]
> @leseb Is CI trying to migrate from Jenkins to github actions?

Not really but who knows :-), I'm experimenting, github actions are much more stable, so if we can move some of the logic there that'd be awesome.

File: .github/workflows/canary-integration-test.yml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  test:
+    runs-on: ubuntu-latest
+    steps:
+    - name: checkout
+      uses: actions/checkout@v2
+
+    - name: setup golang
+      uses: actions/setup-go@v2
+      with:
+        go-version: 1.13
```
Comment (travisn) [inline .github/workflows/canary-integration-test.yml:14]: Can we use go 1.15?

File: .github/workflows/canary-integration-test.yml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    - name: setup minikube
+      uses: opsgang/ga-setup-minikube@v0.1.1
+      with:
+        minikube-version: 1.13.0
+        k8s-version: 1.18.1
```
Comment (travisn) [inline .github/workflows/canary-integration-test.yml:20]: How about the latest k8s 1.19.2?

File: hack/cluster-test-gh-action.yaml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  mon:
+    count: 1
+  dashboard:
+    enabled: false
+  crashCollector:
+    disable: true
+  storage:
+    useAllNodes: true
+    useAllDevices: true
+    deviceFilter: "^DEVICE."
```
Comment (travisn) [inline hack/cluster-test-gh-action.yaml:31]: Why do we need the device filter? What if we just set useAllDevices without the filter?

File: .github/workflows/canary-integration-test.yml [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        kubectl create -f cluster/examples/kubernetes/ceph/nfs-test.yaml
+        kubectl create -f cluster/examples/kubernetes/ceph/toolbox.yaml
+
+    - name: wait for prepare pod
+      run: |
+        timeout 300 sh -c 'until kubectl -n rook-ceph logs -f job/rook-ceph-osd-prepare-minikube; do sleep 5; done'
+
+    - name: wait for ceph to be ready
+      run: |
+        hack/validate_cluster.sh
```
Comment (travisn) [inline .github/workflows/canary-integration-test.yml:68]: If ceph never becomes ready, will the script fail? Or is there a timeout in the script?

File: hack/cluster-test-gh-action.yaml (hunk 5 above)
Comment (travisn) [reply hack/cluster-test-gh-action.yaml:31]: If we still need the filter, what if we just added a new line to the cluster-test.yaml instead of creating this whole new test yaml?
```yaml
# deviceFilter: "^DEVICE."
```

File: hack/validate_cluster.sh [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+function test_demo_rgw {
+  # shellcheck disable=SC2046
+  return $(wait_for_daemon "$EXEC_COMMAND -s | grep -sq 'rgw:'")
+}
+
+function test_demo_mds {
+  echo "Waiting for the MDS to be ready"
+  # NOTE(leseb): metadata server always takes up to 5 sec to run
```
Comment (travisn) [inline hack/validate_cluster.sh:47]: git blame gives us the author, right? 
```suggestion
  # NOTE: metadata server always takes up to 5 sec to run
```

File: hack/validate_cluster.sh [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  # so we first check if the pools exit, from that we assume that
+  # the process will start. We stop waiting after 10 seconds.
+  # shellcheck disable=SC2046
```
Comment (travisn) [inline hack/validate_cluster.sh:50]: should this be uncommented or is it not needed?

File: hack/validate_cluster.sh [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+# MAIN #
+########
+test_demo_mon
+test_demo_mgr
+test_demo_osd
+test_demo_rgw
+test_demo_mds
+test_demo_nfs
+test_demo_rbd_mirror
+
```
Comment (travisn) [inline hack/validate_cluster.sh:95]: How about a check that the csi drivers are running?

File: hack/validate_cluster.sh [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+trap display_status EXIT
+
+#############
+# FUNCTIONS #
+#############
+EXEC_COMMAND="kubectl -n rook-ceph exec $(kubectl get pod -l app=rook-ceph-tools -n rook-ceph -o jsonpath='{.items[0].metadata.name}') -- ceph --connect-timeout 3"
+
+function wait_for_daemon () {
+  timeout=90
```
Comment (travisn) [inline hack/validate_cluster.sh:12]: This timeout may make more sense shorter or longer for some daemons, but we can refine that later.

File: .github/workflows/canary-integration-test.yml (hunk 6 above)
Comment (leseb) [reply .github/workflows/canary-integration-test.yml:68]: Yes, it times out

File: .github/workflows/canary-integration-test.yml (hunk 3 above)
Comment (leseb) [reply .github/workflows/canary-integration-test.yml:14]: We can, but all our builds are on 1.13 so we better stay consistent, right?

File: hack/cluster-test-gh-action.yaml (hunk 5 above)
Comment (leseb) [reply hack/cluster-test-gh-action.yaml:31]: > If we still need the filter, what if we just added a new line to the cluster-test.yaml instead of creating this whole new test yaml?

Then the sed becomes a bit more difficult 👍 
> 
> ```yaml
> # deviceFilter: "^DEVICE."
> ```

Somehow the `useAllDevices` was not working IIRC. Let me try again.

File: hack/validate_cluster.sh (hunk 8 above)
Comment (leseb) [reply hack/validate_cluster.sh:50]: The shellcheck?

File: hack/cluster-test-gh-action.yaml (hunk 5 above)
Comment (leseb) [reply hack/cluster-test-gh-action.yaml:30]: It was failing trying to use a partition... So still using the deviceFilter

**Review by travisn — CHANGES_REQUESTED** [review]

File: hack/validate_cluster.sh (hunk 8 above)
Comment (leseb) [reply hack/validate_cluster.sh:50]: It's a comment to allow the linter to pass
